### ISSUE TITLE
[Form] Validation issue using by_reference false in CollectionType

### ISSUE THREAD

**Issue by diegosardina** [symfony/symfony#32410]
The form doesn't validate properly while using by_reference as false.

Let's assume a trivial one to many association such as product and feature and their related form and collection types. Feature has a non blank field called "name".

Let's add now two elements in the collection and remove the *first one*.

![image](https://user-images.githubusercontent.com/22911186/60758916-48f4a880-a01d-11e9-8ab8-b28310f9c7a8.png)

Form submission is prevented by validation but the validation error doesn't show up.
The current behaviour is shown in the left, while in the right the expected behaviour (with by_reference true).

![image](https://user-images.githubusercontent.com/22911186/60758923-55790100-a01d-11e9-8c44-4a8641c43a74.png)

The issue shown in the Symfony Profiler is pretty straightforward, the collection created (cloned?) has different indexes compared to the one held by the form, then the field doesn't get decorated with its error.

data.features[0].name | This value should not be blank. | null | ConstraintViolation

The one held by the form is data.features[1].name.

As a workaround I regenerated indexes on the form side during the PRE_SUBMIT event.

```
    ->addEventListener(FormEvents::PRE_SUBMIT, function (FormEvent $event) {
        $data = $event->getData();
        if(isset($data['features'])) {
            $data['features'] = array_values($data['features']);
        }
        $event->setData($data);
    })

```

The form should exhibit the same behaviour in both cases (by_reference true or false).

To replicate the issue, it is enough to copy and paste the example in the documentation [here](https://symfony.com/doc/current/form/form_collections.html).

In the Javascript part this sentence:

`
$collectionHolder.data('index', $collectionHolder.find(':input').length);
`

had to be corrected in:

`
$collectionHolder.data('index', $collectionHolder.find(':input').length-1);
`

to work properly, since the former takes into consideration the add button.

**Comment by kl3sk** [comment]
I have the exact same behaviour, @diegosardina hack work for me as should be expected.

As I can see, this is not closed, so not resolved I suppose.

**Comment by xabbuh** [comment]
Can any of you create a small example application that allows to reproduce the issue? I would like to understand in more detail what is going on here.

**Comment by kl3sk** [comment]
@xabbuh I'll prepare a example repo to try to explain more about it.

**Comment by kl3sk** [comment]
here is a [demo repo](https://github.com/kl3sk/symfony-collection-form)

Please advise me is is not self clear.

Don't blame me about the design ;)

**Comment by diegosardina** [comment]
Your demo is ok but not really enough to exhibit the true issue because there is another problem in the documentation. JavaScript code counts in a wrong way because it takes also into account the _add button_ (the [:input selector](https://api.jquery.com/input-selector/) also selects button elements), so an empty collection will start with index 1 instead of 0.

If you are going to _only append_ elements in the collection, it's enough to fix the JavaScript code by replacing:

```
$collectionHolder.data('index', $collectionHolder.find(':input').length);
```
with
```
$collectionHolder.data('index', $collectionHolder.find(':input').length-1);
```

This way you count correctly by starting from 0.

This won't happen with by_reference = true because the collection is reindexed after submit, while with by_reference = false it doesn't.

This is the true issue, that is more visible if you delete elements in the collection.

**Comment by kl3sk** [comment]
Very interesting; I didn't understand that in your previous post.
I totally re-write my own javascript and do not notice this. I'll have a look.

This mean my JS count in the wrong way too and the doc have a typo too.

Let maintainers have look about it. Thank for your hints.

**Comment by kl3sk** [comment]
I can confirm that is just a JavaScript issue, I've made the same mistake on mine, after re-indexing correctly, the expected behaviour occurs.

**Comment by xabbuh** [comment]
Thank you for investigating. I am going to close here then. If you think that the documentation could be improved about this, please open an issue (or even better a PR with a concrete example) in the documentation repository.

**Comment by diegosardina** [comment]
Sorry maybe I wasn't clear enough.

It's not a JavaScript issue, the issue is still there. With a correct JavaScript code the issue just goes unnoticed because there is no problem in appending elements to the collection.

**Comment by xabbuh** [comment]
Sorry, I obviously misunderstood that.

**Comment by kl3sk** [comment]
Yes I misunderstood too, whenever re-indexing correctly some other strange behaviours occurs.

Indeed not a JS issue.

**Comment by diegosardina** [comment]
@kl3sk unfortunately I don't have much time now, could you update your demo with item deletion? 

Deletion of items clearly exhibit the issue.

**Comment by kl3sk** [comment]
I’ll try to find more time to do it. Certainly this evening. 

Indeed this a problem that should be fixed.

**Comment by diegosardina** [comment]
Ok @kl3sk, please accept my PR that fixes some rendering problems. Thank you.

The demo looks good as test case for me.

To replicate the issue: add 2 or more elements and delete the first one, then submit.

**Comment by kl3sk** [comment]
@diegosardina merged

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by chapterjason** [comment]
@xabbuh I think just faced the same issue. The error won't be assigned to the collection but to the form:

<img width="1058" alt="image" src="https://user-images.githubusercontent.com/1337562/111869234-a1e5ab00-897e-11eb-9b07-ed46b0e7e344.png">

I need to use by_reference, I use a many-to-many over an extra entity for an extra position field.
 
Item - One-To-Many <- ItemCategory -> One-To-Many - Category

**Comment by kl3sk** [comment]
I maybe misunderstood your question, but your error looks like a assert like the ‘@Assert\Count‘.

**Comment by chapterjason** [comment]
@kl3sk yes, but on the property `xxxCategories`. I Expect, like email, that the error will be mapped to the `xxxCategories` property but it will mapped to the root form. Currently working on a reproducer.

**Comment by chapterjason** [comment]
@xabbuh @kl3sk I managed to create a [reproducer](https://github.com/chapterjason/symfony_issue_32410), this also happen even without rendering the form or use client side logic. I use `FormInterface::submit`  and the error is mapped wrong.

// **Edit**
I debugged a bit and set `error_bubbling` to `false`.
This will cause that the error will be added to the CollectionType form field.

**Comment by rik702** [comment]
I've also got this issue and have been experimenting with a possible fix.  I'm not all that familiar with the internals of Symfony though so comments are requested.  
Within the function "writeCollection" in "PropertyAccessor" there is a loop:
```
      foreach ($collection as $item) {
            if (!$previousValue || !\in_array($item, $previousValue, true)) {
                $zval[self::VALUE]->$addMethodName($item);
            }
        }
```
I have updated this to:
```
        foreach ($collection as $key => $item) {
            if (!$previousValue || !\in_array($item, $previousValue, true)) {
                $zval[self::VALUE]->$addMethodName($item, collectionKeyForRecreatingSubmittedFormKeyInCaseOfValidationError: $key);
            }
        }
```
ie added the $key from $collection as an (optional) parameter to the add*() function on the object.

I can then use this key in the add*() function in the Entity to index the collection to match the indices passed up from the form submit.  This in turn allows the renderer to match the correct index on the page with an error message.

PS.   
Line 708 in the "ReflectionExtractor" class, in the function "findAdderAndRemover" causes an exception (below)
`[$addMethodFound, $addMethodAccessibleErrors] = $this->isMethodAccessible($reflClass, $addMethod, 1);`

if the $key parameter is not optional in the add*() function.

PPS
After further searching around the Symfony PRs have found 2 addressing the same problem #22867 and #39438 .  I have also created a reproducer with an amendment to the above suggestion now including a Named Argument for the key.  The reproducer is [https://github.com/rik702/symfony-issue-32410](https://github.com/rik702/symfony-issue-32410)

PPPS
OK, after reading all PRs and issues I could find on the matter (there are a lot going back some years) I decided a workaround is the best solution.   As there is a large amount of convention needed in the web app code anyway, it seems the sensible solution.  I've ended up creating a subscriber that reindexes the Doctrine Collections that are used in a form.  The solution is in the reproducer [https://github.com/rik702/symfony-issue-32410/commit/fab45a9a3850bbb3771625d408a50c5e180a2d1d](https://github.com/rik702/symfony-issue-32410/commit/fab45a9a3850bbb3771625d408a50c5e180a2d1d) .  If anyone can suggest a better way of reindexing the collection, even better...
